### PR TITLE
Delayed param

### PR DESCRIPTION
**Pull request by ordabayevy**
Addresses #533

### CODE REVIEW

File: funsor/minipyro.py [hunk 1]
```diff
@@ -21,9 +21,9 @@
 
 import torch
 from pyro.distributions import validation_enabled
-from pyro.optim.clipped_adam import ClippedAdam as _ClippedAdam
 
 import funsor
+from funsor.adam import Adam  # noqa: F401
```
Comment (ordabayevy) [inline funsor/minipyro.py:26]: for pyroapi compatibility

File: funsor/minipyro.py [hunk 2]
```diff
@@ -306,11 +330,13 @@ def param(
...
-        if event_dim is None:
-            event_dim = init_value.dim()
-        output = funsor.Reals[init_value.shape[init_value.dim() - event_dim :]]
+    # infer output
+    value = init_value
+    if value is None:
+        value, _ = PARAM_STORE[name]
+    if event_dim is None:
+        event_dim = value.dim()
+    output = funsor.Reals[value.shape[value.dim() - event_dim :]]
```
Comment (ordabayevy) [inline funsor/minipyro.py:339]: infer `output` when `pyro.param` was already defined elsewhere

File: funsor/minipyro.py [hunk 3]
```diff
@@ -408,24 +400,33 @@ def __init__(self, model, guide, optim, loss):
...
+            name = name.replace("_unconstrained", "")
+            value = value.data
+            old_value, constraint = PARAM_STORE[name]
+            if old_value is not value:
+                old_value.data.copy_(value)
+        return result
+
+    def step(self, *args, **kwargs):
+        self.optim.num_steps = 1
+        return self.run(*args, **kwargs)
```
Comment (ordabayevy) [inline funsor/minipyro.py:429]: for compatibility with `SVI` interface

File: funsor/minipyro.py [hunk 4]
```diff
@@ -408,24 +400,33 @@ def __init__(self, model, guide, optim, loss):
...
-        funsor.to_data(loss).backward()
-        # Grab all the parameters from the trace.
-        params = [site["value"].data.unconstrained() for site in param_capture.values()]
-        # Take a step w.r.t. each parameter in params.
-        self.optim(params)
-        # Zero out the gradients so that they don't accumulate.
-        for p in params:
-            p.grad = torch.zeros_like(p.grad)
-        return loss.item()
+        with funsor.terms.lazy:
```
Comment (ordabayevy) [inline funsor/minipyro.py:407]: `lazy` interpretation is needed here to make sure that `funsor.Integrate` is not eagerly expanded in `Expectation`

**Comment by ordabayevy** [comment]
`examples/minipyro.py` is currently failing with jit. My guess is that jitting needs to be baked into `funsor.adam.Adam`, I will think more about this.

**Comment by ordabayevy** [comment]
Am I right that when using `funsor.adam.Adam` the function that needs to be jit traced is the loss function below (`Subs` funsor) @fritzo @eb8680 ? If yes, then it first needs to be converted to  a function with positional arguments?

```py
step_loss = loss(**{k: v[...] for k, v in params.items()}).data
```

File: funsor/minipyro.py (hunk 3 above)
Comment (fritzo) [reply funsor/minipyro.py:429]: Hmm, let's think about alternative workarounds... One issue here is that the Adam optimizer statistics would not be persisted across svi steps.

One option is simply to change pyroapi's SVI interface to look for either `.run()` or if missing fall back to `.step()`.

**Comment by fritzo** [comment]
> ...failing with jit. My guess is that jitting needs to be baked into `funsor.adam.Adam`

I think you're right, but let's discuss.  That's a little different from Pyro where jit is baked into ELBO subclasses.
